src/Functions/array/arrayNorm.cpp

```diff
@@ -159,46 +173,35 @@ class FunctionArrayNorm : public IFunction
         }
     }
 
-    template <typename MatrixType, typename DataType>
-    requires std::is_same_v<MatrixType, DataType>
-    void fillVectors(std::vector<Eigen::VectorX<MatrixType>> & vec, const ColumnArray & array) const
+    template <typename ResultType, typename ArgumentType>
+    static ColumnPtr executeWithTypes(const ColumnArray & array, size_t input_rows_count)
     {
-        const auto & data = typeid_cast<const ColumnVector<DataType> &>(array.getData()).getData();
+        const auto & data = typeid_cast<const ColumnVector<ArgumentType> &>(array.getData()).getData();
         const auto & offsets = array.getOffsets();
-        vec.reserve(offsets.size());
-        ColumnArray::Offset prev = 0;
-        for (auto off : offsets)
-        {
-            vec.emplace_back(Eigen::Map<const Eigen::VectorX<MatrixType>>(data.data() + prev, off - prev));
-            prev = off;
-        }
-    }
 
-    template <typename MatrixType, typename DataType>
-    void fillVectors(std::vector<Eigen::VectorX<MatrixType>> & vec, const ColumnArray & array) const
-    {
-        const auto & data = typeid_cast<const ColumnVector<DataType> &>(array.getData()).getData();
-        const auto & offsets = array.getOffsets();
-        vec.reserve(offsets.size());
+        auto result_col = ColumnVector<ResultType>::create(input_rows_count);
+        auto & result_data = result_col->getData();
 
         ColumnArray::Offset prev = 0;
+        size_t row = 0;
         for (auto off : offsets)
         {
-            Eigen::VectorX<MatrixType> mat(off - prev);
-            for (ColumnArray::Offset row = 0; row + prev < off; ++row)
+            Float64 result = 0;
+            for (; prev < off; ++prev)
```

Comment:
maybe move this loop inside `accumulate`? this way explicit `finalize` would no longer be needed and the chances that the compiler will optimize the loop would be at least not less.

Reply:
Actually I looked at disassembly and I saw that the compiler did a decent job vectorizing the loop with SSE:

L2Norm:

```
 14.50 │120:   movsd   (%rcx,%rbp,8),%xmm1                                                                                                                     ▒
  0.08 │       movsd   0x8(%rcx,%rbp,8),%xmm2                                                                                                                  ▒
  0.27 │       mulsd   %xmm1,%xmm1                                                                                                                             ▒
  4.24 │       addsd   %xmm0,%xmm1                                                                                                                             ▒
  3.80 │       mulsd   %xmm2,%xmm2                                                                                                                             ▒
  0.88 │       addsd   %xmm1,%xmm2                                                                                                                             ▒
 11.89 │       movsd   0x10(%rcx,%rbp,8),%xmm1                                                                                                                 ◆
  3.67 │       mulsd   %xmm1,%xmm1                                                                                                                             ▒
  4.28 │       addsd   %xmm2,%xmm1                                                                                                                             ▒
       │       movsd   0x18(%rcx,%rbp,8),%xmm0                                                                                                                 ▒
  0.00 │       mulsd   %xmm0,%xmm0                                                                                                                             ▒
  4.72 │       addsd   %xmm1,%xmm0                                                                                                                             ▒
  3.69 │       add     $0x4,%rbp                                                                                                                               ▒
       │       cmp     %rbp,%rax                                                                                                                               ▒
  3.29 │     ↑ jne     120   
```

L2Distance:

```
  0.01 │260:   movsd   (%rsi,%rdi,8),%xmm1                                                                                                                     ▒
 23.01 │       subsd   -0x8(%rcx,%rdi,8),%xmm1                                                                                                                 ▒
  0.62 │       mulsd   %xmm1,%xmm1                                                                                                                             ▒
 14.57 │       addsd   %xmm0,%xmm1                                                                                                                             ▒
  0.01 │       movsd   0x8(%rsi,%rdi,8),%xmm0                                                                                                                  ▒
  0.02 │       subsd   (%rcx,%rdi,8),%xmm0                                                                                                                     ▒
  0.02 │       mulsd   %xmm0,%xmm0                                                                                                                             ◆
 14.31 │       addsd   %xmm1,%xmm0                                                                                                                             ▒
  0.01 │       add     $0x2,%rdi                                                                                                                               ▒
       │       cmp     %rdi,%rdx                                                                                                                               ▒
  9.99 │     ↑ jne     260      

```

cosineDistance:

```
  0.00 │2a0:   movsd   (%rcx,%rdi,8),%xmm2
  0.01 │       movapd  %xmm2,%xmm3
  2.14 │       mulsd   %xmm2,%xmm2
  8.15 │       addsd   %xmm0,%xmm2
  0.00 │       movsd   0x8(%rcx,%rdi,8),%xmm0
  9.55 │       movsd   -0x8(%rdx,%rdi,8),%xmm4
  2.58 │       mulsd   %xmm4,%xmm3
  8.03 │       mulsd   %xmm4,%xmm4
  0.29 │       addsd   %xmm1,%xmm4
  0.00 │       movsd   (%rdx,%rdi,8),%xmm1
  2.13 │       addsd   %xmm7,%xmm3
  8.11 │       movapd  %xmm0,%xmm7
  0.12 │       mulsd   %xmm1,%xmm7
  0.48 │       addsd   %xmm3,%xmm7
  2.04 │       mulsd   %xmm0,%xmm0
  7.95 │       addsd   %xmm2,%xmm0
  0.00 │       mulsd   %xmm1,%xmm1
  0.21 │       addsd   %xmm4,%xmm1
  2.08 │       add     $0x2,%rdi
  2.05 │       cmp     %rdi,%rsi
  8.07 │     ↑ jne     2a0

```

Reply:
As for moving the loop inside accumulate() I see you point but in case of distance functions there are 2 slightly different kinds of loops:
1. for 2 non-const column arguments
```
        for (auto off : offsets_x)
        {
            typename Kernel::template State<Float64> state;
            for (; prev < off; ++prev)
            {
                Kernel::accumulate(state, data_x[prev], data_y[prev]);
            }
            result_data[row] = Kernel::finalize(state);
            row++;
        }
```
2. for 1 const and 1 non-const argument
```
        for (auto off : offsets_y)
        {
            typename Kernel::template State<Float64> state;
            for (size_t i = 0; prev < off; ++i, ++prev)
            {
                Kernel::accumulate(state, data_x[i], data_y[prev]);
            }
            result_data[row] = Kernel::finalize(state);
            row++;
        }
```
In these cases it will require 2 different implementations.
So I think as long as the compiler is able to inline accumulate() it's cleaner to keep it like this.